mixer/pkg/adapter/opencensus/trace.go

```diff
@@ -0,0 +1,253 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package opencensus contains support code for writing adapters that use OpenCensus.
+package opencensus
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.opencensus.io/plugin/ochttp"
+	"go.opencensus.io/trace"
+
+	"istio.io/istio/mixer/template/tracespan"
+)
+
+const attrHTTPStatusCode = "http.status_code"
+
+// Handler implements tracespan.Handler using an OpenCensus TraceExporter.
+type Handler struct {
+	exporters sync.Map
+	// NewExporter creates OpenCensus exporters where generated spans will be exported.
+	NewExporter NewExporterFunc
+	// Sampler to be applied to each span.
+	Sampler trace.Sampler
+	// CloseFunc will be called when this handler is closed. Optional.
+	CloseFunc func() error
+}
+
+type exporterKey struct {
+	name     string
+	endpoint string
+}
+
+var _ tracespan.Handler = (*Handler)(nil) // type assertion
+
+// NewExporterFunc is a callback to create new OpenCensus exporters for a given
+// workload name and endpoint.
+//
+// The format of endpoint will be IP_ADDRESS:PORT. name and endpoint refer to
+// the logical owning workload of this span. For server spans, the values are the
+// destinationName and destinationIp, for clients the sourceName and sourceIp.
+type NewExporterFunc func(name string, endpoint string) trace.Exporter
+
+// NewTraceHandler returns a new tracespan adapter that sends spans to the provided
+// exporter.
+func NewTraceHandler(newExporter NewExporterFunc, sampler trace.Sampler) *Handler {
+	return &Handler{
+		NewExporter: newExporter,
+		Sampler:     sampler,
+	}
+}
+
+// HandleTraceSpan transforms tracespan template instances into OpenCensus spans
+// and sends them to the configured OpenCensus exporter.
+func (h *Handler) HandleTraceSpan(_ context.Context, values []*tracespan.Instance) (retErr error) {
+	if h.Sampler == nil {
+		// Tracing is not configured.
+		return nil
+	}
+
+	numExported := 0
+	for _, val := range values {
+		parentContext, ok := ExtractParentContext(val.TraceId, val.ParentSpanId)
+		if !ok {
+			continue
+		}
+		spanContext, ok := ExtractSpanContext(val.SpanId, parentContext)
+		if !ok {
+			continue
+		}
+
+		decision := h.Sampler(trace.SamplingParameters{
+			ParentContext:   parentContext,
+			TraceID:         spanContext.TraceID,
+			SpanID:          spanContext.SpanID,
+			Name:            val.SpanName,
+			HasRemoteParent: true,
+		})
+
+		if !decision.Sample {
+			continue
+		}
+		spanContext.TraceOptions = trace.TraceOptions(1 /*sampled*/)
+
+		span := buildSpanData(val, parentContext, spanContext)
+		h.exporter(val).ExportSpan(span)
```

Comment:
does `ExportSpan` block at all?  Should we consider moving this to a goroutine to avoid potential contention?

Reply:
actually, just read the spec for that interface... seems fine.  Nevermind.